2471: Edit accepting store frontend

f1sh1918 opened the pull request:
### Short Description

In this pr a menu item "bearbeiten" and a "Neu hinzufügen" Button for stores is added. Clicking that a modal opens with a form that let you edit existing store data or create new store data. Mind this form is not connected to the backend.
I also added the required form validations and connected the geodata service to check if a typed address is valid.

During that implementation i refactored some FormFields to be able to reuse them and unify the behavior (validation f.e.)
Since this pr is even big, i will create some follow up tickets, when we discussed the topic.
Some types may have to be adjusted slightly when the backend function is ready

###  Follow-ups: 
- Use `CustomSelect` for all Selects to unify them include same error handling
- rename `CardTextField` to `CustomTextField`
- use `CustomTextField` for multiline fields
- check if some TextFields that use custom onBlur can be converted to `CustomTextFields`
- improve geolocation coordinates (to be discussed)
- split street and houseNr in the, has to be discussed -> separate issue

### Proposed Changes

<!-- Describe this PR in more detail. -->

- add ManageStoreDialog that contains the StoreForms with different sections
- add form validation as requested
- connect to geoService to fetch coordinates

### Side Effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->

- none

### Testing

1. Login as ProjectStoreManager
2. Click the Menu Button of the table and "Bearbeiten", test the validations
3. Do the same with "Neuen Store hinzufügen"

### Resolved Issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #2471

f1sh1918 commented on administration/src/util/helper.ts:
```diff
@@ -9,6 +9,7 @@ import PlainDate from './PlainDate'
 export const isStagingEnvironment = (): boolean => !!window.location.hostname.match(/staging./)
 export const isProductionEnvironment = (): boolean =>
   process.env.REACT_APP_IS_PRODUCTION === 'true' && !isStagingEnvironment()
+export const isDevelopmentEnvironment = (): boolean => process.env.REACT_APP_IS_PRODUCTION === 'false'
```
hm i think f.e. checking f.e. for `localhost` might not be a clear check because you can also access via network ip if you use you mobile device to test.
So afaics we could only check if its staging or production via projectConfig and take development as a fallback if the other two checks don't match. I think this would be quite risky

f1sh1918 commented on administration/src/routes/stores/components/form/validation.tsx:
```diff
@@ -0,0 +1,197 @@
+const EMAIL_REGEX = /^[a-zA-Z0-9._%+-]+(?<!\.)@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*\.[a-zA-Z]+$/
+
+export type FormValidation = {
+  invalid: boolean
```
i understand this opinion but when we use the validation in form components, we have `errorMessage` or `showError` and then we would reverse it there. I think its in the usage in the component more intuitive

andrew8er commented:
> No because of mandatory fields left empty, we also show error if it was focused and unfocused.
For me this is also interaction.

Makes sense, but it can be both.

f1sh1918 commented on administration/src/routes/stores/components/form/DescriptionSection.tsx:
i fixed the label and adjust it to show `0` if there is more than the accepted amount of characters

f1sh1918 commented:
Thx for this detailed review.
I think i addressed everything and resolved the very obvious things.
Please check the rest that is still open.

There are some general things like error handling in forms which should be discussed and adjusted in a separate issue, since the issue itself is big enough by far.
We decided not to show the error immediately but when the user shows that he/she is ready and leaves the field. This can be challenged but there should be a general discussion with ui/ux about that

andrew8er commented on administration/src/routes/region/util/storeGeoDataService.ts:
```diff
@@ -20,12 +20,12 @@ const handleStoreWithMissingLocationInformation = (
   geoServiceUrl.searchParams.append('countrycodes', 'de')
   geoServiceUrl.searchParams.append('city', location)
-  geoServiceUrl.searchParams.append('street', `${houseNr} ${street}`)
+  geoServiceUrl.searchParams.append('street', `${street}`)
```
Ah makes sense. I was a little confused, because [here](https://github.com/digitalfabrik/entitlementcard/pull/2707/files#diff-5d0c64a36e47a91e51e958b8b1c32a196dbbce0563a4aed48cca0b2ff2d9dfc3R37) the query function is called with a separate house number parameter.

andrew8er commented on administration/src/routes/stores/components/form/validation.tsx:
```diff
@@ -0,0 +1,197 @@
+  DESCRIPTION: { min: 0, max: DESCRIPTION_MAX_CHARS },
+}
+
+const ADDRESS_REGEX = /^[a-zA-ZäöüÄÖÜß0-9\s.-]+$/
```
Works 👍

bahaaTuffaha reviewed (approved):
Great work ✅
